pex/toml.py

```diff
@@ -3,61 +3,232 @@
 
 from __future__ import absolute_import
 
+import re
 import sys
+from collections import OrderedDict
+from datetime import date, datetime, time
 from io import BytesIO
 
+from pex.compatibility import string
 from pex.typing import TYPE_CHECKING, cast
 
 if TYPE_CHECKING:
-    from typing import Any, Dict, Union
+    from typing import IO, Any, Dict, List, Text, Tuple, Union
 
 TOMLI_SUPPORTED = sys.version_info[:2] >= (3, 7)
 
 
 if not TOMLI_SUPPORTED:
     from pex.third_party.toml import TomlDecodeError as _TomlDecodeError
-    from pex.third_party.toml import dump as _dump
-    from pex.third_party.toml import dumps as _dumps
     from pex.third_party.toml import load as _load
     from pex.third_party.toml import loads as _loads
 
     def load(source):
-        # type: (Union[str, BytesIO]) -> Dict[str, Any]
+        # type: (Union[str, IO[bytes]]) -> Dict[str, Any]
         if isinstance(source, str):
             return cast("Dict[str, Any]", _load(source))
         else:
             return cast("Dict[str, Any]", _loads(source.read().decode("utf-8")))
 
-    def dump(
-        data,  # type: Dict[str, Any]
-        fp,  # type: BytesIO
-    ):
-        # type: (...) -> None
-        fp.write(_dumps(data).encode("utf-8"))
-
 else:
     from pex.third_party.tomli import TOMLDecodeError as _TomlDecodeError
     from pex.third_party.tomli import load as _load
     from pex.third_party.tomli import loads as _loads
-    from pex.third_party.tomli_w import dump as _dump
-    from pex.third_party.tomli_w import dumps as _dumps
 
     def load(source):
-        # type: (Union[str, BytesIO]) -> Dict[str, Any]
+        # type: (Union[str, IO[bytes]]) -> Dict[str, Any]
         if isinstance(source, str):
             with open(source, "rb") as fp:
                 return cast("Dict[str, Any]", _load(fp))
         else:
             return cast("Dict[str, Any]", _load(source))
 
-    def dump(
-        data,  # type: Dict[str, Any]
-        fp,  # type: BytesIO
-    ):
-        # type: (...) -> None
-        _dump(data, fp)
-
 
 loads = _loads
 TomlDecodeError = _TomlDecodeError
-dumps = _dumps
+
+
+class InlineTable(OrderedDict):
```

Comment:
N.B.: All this new toml writing code is driven by the need to control when an inline table is emitted to match spec examples / uv / general expectations of the output format, which is both pretty unreadable with inline tables and without, but where the public preference seems to have been demonstrated as having inline tables to keep a package's info grouped. Neither the toml or tomli-w libraries vendored by Pex allow control over this. The tomlkit library by Poetry folks does, but is Python>=3.8 only.

I still find it insane that PyPA adopts toml for standards when officially supported Pythons don't support toml (Python >=3.9,<3.11) and, worse for tool authors, no Python versions support writing toml. The only options are non-PyPA owned third party libraries or roll you own. Not cool!